rasa/core/policies/ensemble.py

```diff
@@ -91,590 +218,316 @@ def _check_for_important_policies(self) -> None:
+                    f"different priorities.",
+                    docs=DOCS_URL_POLICIES,
+                )
 
-        # also check if confidence is 0, than it cannot be count as prediction
-        return not is_memorized or max_confidence == 0.0
+    def combine_predictions(
+        self, predictions: List[PolicyPrediction2], tracker: DialogueStateTracker
+    ) -> PolicyPrediction2:
+        """Derives a single prediction from the given list of predictions.
 
-    def _pick_best_policy(
-        self, predictions: Dict[Text, PolicyPrediction]
-    ) -> PolicyPrediction:
-        """Picks the best policy prediction based on probabilities and policy priority.
+        Note that you might get unexpected results if the priorities are non-unique.
+        Moreover, the order of events in the result is determined by the order of the
+        predictions passed to this method.
 
         Args:
-            predictions: the dictionary containing policy name as keys
-                         and predictions as values
+            predictions: a list of policy predictions that include "probabilities"
+              which are non-negative but *do not* necessarily up to 1
+            tracker: dialogue state tracker holding the state of the conversation
 
         Returns:
-            The best prediction.
+            The "best" prediction.
         """
-        best_confidence = (-1, -1)
-        best_policy_name = None
+        if not predictions:
+            raise InvalidConfigException(
+                "Expected at least one prediction. Please check your model "
+                "configuration."
+            )
+        # Reminder: If just a single policy is given, we do *not* just return it because
+        # it is expected that the final prediction contains mandatory and optional
+        # events in the `events` attribute and no optional events.
 
-        # different type of predictions have different priorities
-        # No user predictions overrule all other predictions.
-        is_no_user_prediction = any(
-            prediction.is_no_user_prediction for prediction in predictions.values()
-        )
-        # End-to-end predictions overrule all other predictions based on user input.
-        is_end_to_end_prediction = any(
-            prediction.is_end_to_end_prediction for prediction in predictions.values()
+        best_idx = self._choose_best_prediction(predictions)
+        events = self._choose_events(
+            predictions=predictions, best_idx=best_idx, tracker=tracker
         )
 
-        policy_events = []
-        for policy_name, prediction in predictions.items():
-            policy_events += prediction.events
+        # make a copy and swap out the events - and drop action metadata
+        # (TODO: what is action meta data and why was it ignored?)
+        final = copy.copy(predictions[best_idx])
+        final.events = events
+        final.optional_events = []
+        final.action_metadata = None
```

Comment:
`UnexpecTED` is using this and it should definitely not be ignored 🙈

Reply:
great catch 👍🏻

Reply:
Sorry that comment is confusing. The old implementation instantiates a new prediction with everything except the `action_metadata`. So this new version is the same as before - I was just wondering if there *should be* something.  ....So shouldn't this be `None`?

Reply:
Exactly. The `action_metadata` shouldn't be `None`. Instead the action metadata should be passed over to the final `PolicyPrediction`

Reply:
fixed this in the reverted version